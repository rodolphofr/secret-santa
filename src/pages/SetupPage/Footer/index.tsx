import { useNavigate } from "react-router-dom";
import "./Footer.css";

type Props = {
  participants: string[];
};

const MINIMUM_PLAYERS_COUNT = 3;

const Footer = ({ participants }: Props) => {
  const navigate = useNavigate();
  const canSort = participants.length >= MINIMUM_PLAYERS_COUNT;

  return (
    <footer>
      <button
        className="actionButton"
        disabled={!canSort}
        onClick={() => navigate("/draw")}
      >
        Iniciar bricadeira!
      </button>
      <img src="/images/bags.png" alt="imagem de sacolas" />
    </footer>
  );
};

export default Footer;
