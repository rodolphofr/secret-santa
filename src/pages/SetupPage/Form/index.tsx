import { useFormik } from "formik";
import { useEffect, useRef } from "react";
import * as Yup from "yup";
import "./Form.css";

type Props = {
  participants: string[];
  onInsert: (participant: string) => void;
};

type Values = {
  participant: string;
}

const Form = ({ participants, onInsert }: Props) => {
  const inputRef = useRef<HTMLInputElement>(null);

  const {
    errors,
    values,
    touched,
    handleSubmit,
    handleChange,
    setFieldError,
  } = useFormik<Values>({
    initialValues: { participant: "" },
    validateOnBlur: false,
    validateOnMount: false,
    validateOnChange: false,
    validationSchema: Yup.object().shape({
      participant: Yup
        .string()
        .trim()
        .required("Adicione o nome do participant")
        .test(
          "check-duplicate-names",
          "participant jÃ¡ adicionado",
          (newParticipant) => {
            return !!newParticipant && !participants.includes(newParticipant);
          }
        )
    }),
    onSubmit: ({ participant }, { resetForm }) => {
      onInsert(participant);
      resetForm();
      inputRef.current?.focus();
    }
  });

  useEffect(() => {
    let timeoutId: NodeJS.Timeout;

    if (errors.participant)
      timeoutId = setTimeout(() => setFieldError("participant", undefined), 3000);

    return () => {
      clearTimeout(timeoutId);
    };
  }, [errors, setFieldError]);

  return (
    <form onSubmit={handleSubmit}>
      <div className="wrapperFields">
        <div className="group">
          <input
            ref={inputRef}
            type="text"
            id="participant"
            name="participant"
            data-testid="participant"
            autoComplete="off"
            placeholder="Insira os nomes dos participantes"
            onChange={handleChange}
            value={values.participant}
          />
          <button
            type="submit"
            data-testid="addButton"
            disabled={!values.participant}
          >
            Adicionar
          </button>
        </div>

        {errors.participant && touched.participant && (
          <p className="error" role="alert">{errors.participant}</p>
        )}
      </div>
    </form>
  );
};

export default Form;
