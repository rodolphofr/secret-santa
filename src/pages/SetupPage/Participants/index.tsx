import "./Participants.css";

type Props = {
  participants: string[];
  onDeleteItem: (participante: string) => void;
};

const Participants = ({ participants, onDeleteItem }: Props) => (
  <ul>
    {participants.map(participant => (
      <li key={participant} onClick={() => onDeleteItem(participant)}>
        {participant}
      </li>
    ))}
  </ul>
);

export default Participants;
