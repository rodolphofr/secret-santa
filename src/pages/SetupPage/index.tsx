import Form from './Form'
import Participants from './Participants'
import Footer from './Footer';

type Props = {
  participants: string[];
  addParticipant: (participante: string) => void;
  deleteParticipant: (participante: string) => void;
};

const SetupPage = ({
  participants,
  addParticipant,
  deleteParticipant,
}: Props) => (
  <>
    <h2>Vamos come√ßar!</h2>
    <Form
      participants={participants}
      onInsert={addParticipant}
    />
    <Participants
      participants={participants}
      onDeleteItem={deleteParticipant}
    />
    <Footer participants={participants} />
  </>
);

export default SetupPage;
