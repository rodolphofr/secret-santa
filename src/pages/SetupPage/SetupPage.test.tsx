import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import SetupPage from ".";

test("Pagina de configuracao deve ser renderizada", () => {
  const { container } = render(
    <MemoryRouter>
      <SetupPage
        participants={[]}
        addParticipant={jest.fn()}
        deleteParticipant={jest.fn()}
      />
    </MemoryRouter>
  );

  expect(container).toMatchSnapshot();
});
