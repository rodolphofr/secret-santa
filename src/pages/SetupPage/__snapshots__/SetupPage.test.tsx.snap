// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagina de configuracao deve ser renderizada 1`] = `
<div>
  <h2>
    Vamos come√ßar!
  </h2>
  <form>
    <div
      class="wrapperFields"
    >
      <div
        class="group"
      >
        <input
          autocomplete="off"
          data-testid="participant"
          id="participant"
          name="participant"
          placeholder="Insira os nomes dos participantes"
          type="text"
          value=""
        />
        <button
          data-testid="addButton"
          disabled=""
          type="submit"
        >
          Adicionar
        </button>
      </div>
    </div>
  </form>
  <ul />
  <footer>
    <button
      class="actionButton"
      disabled=""
    >
      Iniciar bricadeira!
    </button>
    <img
      alt="imagem de sacolas"
      src="/images/bags.png"
    />
  </footer>
</div>
`;
